### PR TITLE
Topology-aware scheduling

### PR DESCRIPTION
**Pull request by aaronlehmann**
This adds support for placement preferences in Swarm services (design doc: https://github.com/docker/swarmkit/pull/1473, swarmkit PR: https://github.com/docker/swarmkit/pull/1512). This feature allows services to be balanced over nodes based on a particular user-defined property, such as which datacenter or rack they are located in.

- Convert `PlacementPreferences` between GRPC API and HTTP API
- Add `--placement-pref-add` and `--placement-pref-rm` to CLI
- Add support for placement preferences in `service inspect --pretty`
- Add integration test

cc @aluzzardi @dongluochen @dnephin

### CODE REVIEW

File: cli/command/formatter/service.go
Comment (aaronlehmann) [inline cli/command/formatter/service.go]: Note: `Constraints` was misspelled here, and the extra `-` in the template on the line above caused the lines of text to get merged together. These problems are fixed here. Unfortunately it's not the first time I've noticed problems like this in the template. We should review it carefully.

**Review by dnephin — COMMENTED** [review]
Design LGTM

**Comment by aaronlehmann** [comment]
swarmkit PR was merged. No longer WIP.

Changed the command line syntax from `spread,...` to `spread=...` as suggested in https://github.com/docker/swarmkit/pull/1512#issuecomment-278162217

**Comment by vdemeester** [comment]
@aaronlehmann some test failures

```
22:47:48 # github.com/docker/docker/cli/command/service
22:47:48 cli/command/service/update_test.go:63: undefined: updatePlacement
```

**Review by vdemeester — COMMENTED** [review]
Design LGTM

**Comment by aaronlehmann** [comment]
Fixed, sorry about that.

**Review by dnephin — APPROVED** [review]
LGTM

**Comment by aaronlehmann** [comment]
Rebased.

File: cli/command/service/opts.go
Comment (vdemeester) [inline cli/command/service/opts.go]: missing a `"` 👼

**Review by vdemeester — APPROVED** [review]
LGTM 🐸
/cc @thaJeztah @vieux

File: cli/command/service/opts.go
Comment (aaronlehmann) [inline cli/command/service/opts.go]: Thanks, fixed

**Review by cpuguy83 — APPROVED** [review]
LGTM

**Comment by cpuguy83** [comment]
Needs some docs.

**Comment by aaronlehmann** [comment]
Happy to see that this went through the code review process relatively quickly and easily. Now that the code changes have been finalized, I'll work on some documentation in the next day or two.

**Comment by aaronlehmann** [comment]
Added documentation in `service_create.md` and opened https://github.com/docker/docker.github.io/pull/1647 to reference this in the Docker Services guide. PTAL.

File: docs/reference/commandline/service_create.md
Comment (vdemeester) [inline docs/reference/commandline/service_create.md]: Hum, we tend to not have the `*-add` in the `create` subcommand (would be `--placement-pref` instead)

**Review by vdemeester — COMMENTED** [review]
/cc @thaJeztah 👼:pray:

**Comment by aaronlehmann** [comment]
> Hum, we tend to not have the *-add in the create subcommand (would be --placement-pref instead)
> 

Yeah, I discussed this in the swarmkit PR. The issue is I think it would be extremely confusing for `--placement-pref` to be so sensitive to ordering. Reordering two or more of them would lead to totally different behavior. This doesn't feel right to me.

I think it's slightly better (but still not perfect) if the option is called `--placement-pref-add`. Since this is a verb, it at least hints that the flag is *doing* something instead of just providing information.

Originally I wanted to call it `--append-placement-pref`, but that's even more inconsistent.

I'm prepared to back down on this and switch to `--placement-pref`, but as I said it still doesn't feel right to me.

**Comment by vdemeester** [comment]
@aaronlehmann Ohh I didn't think of that (ordering). Hum I guess it makes sense (cc @thaJeztah)

**Comment by thaJeztah** [comment]
Do we have other options that depend on ordering? I agree that `-add` is inconsistent from a UX perspective; _almost_ tempted to say if ordering is important, it should be a comma-separated list.

If I understand correctly, these services have a very different placement preference;

```
docker service create \
  --name=serviceA \
  --placement-pref-add=one \
  --placement-pref-add=two \
  --placement-pref-add=three \
  foobar
```

```
docker service create \
  --name=serviceB \
  --placement-pref-add=three \
  --placement-pref-add=two \
  --placement-pref-add=one \
  foobar
```

How do I update my service from "serviceA" to "serviceB"? This would probably become;

```
docker service update \
  --placement-pref-rm=one \
  --placement-pref-rm=two \
  --placement-pref-rm=three \
  --placement-pref-add=three \
  --placement-pref-add=two \
  --placement-pref-add=one \
  serviceA
```

That's a bit awkward.

If it would be a comma-separated list, this would be;

```
docker service create \
  --name=serviceA \
  --placement-pref=one,two,three \
  foobar
```

```
docker service update \
  --placement-pref=three,two,one \
  serviceA
```

That's really just a quick thought, so putting it here for discussion

**Comment by aaronlehmann** [comment]
Thanks for the thoughtful response. This was my initial instinct, but using a comma-separated list has some serious disadvantages as well. See the discussion thread starting at https://github.com/docker/swarmkit/pull/1512#issuecomment-273923962 where this was discussed, and the alternative of using a repeated flag was encouraged.

**Comment by dnephin** [comment]
Looks like this needs a rebase.

> Do we have other options that depend on ordering? 

Yes we do, but Aaron noted that a change in order for most of those options is less significant than in this case.

**Comment by aaronlehmann** [comment]
Rebased

**Comment by aluzzardi** [comment]
I do agree with UX consistency as well...

`--placement-pref` would have the same behavior as `--placement-pref-add` during `service create`, it's just called differently...

**Comment by aaronlehmann** [comment]
Then is there a consensus around calling it `--placement-pref` in `service create` and `--placement-pref-add` in `service update`?

It's not my favorite outcome, but I don't see another way to move forward.

**Comment by dnephin** [comment]
Yes, I think that is the most consistent naming

**Comment by aaronlehmann** [comment]
Okay, I've updated this to use `--placement-pref` for `service create`.

**Comment by vieux** [comment]
that's very cool, LGTM!

**Comment by aaronlehmann** [comment]
@thaJeztah: Are you happy with the latest change to the CLI flags?

**Review by thaJeztah — APPROVED** [review]
LGTM, new names for the flags work 👍 

I see there's no man page yet for these commands, so I won't hold up this PR because of that (it's something to be addressed in general)

**Comment by Shadowsvoices** [comment]
Is this supported when deploying services via docker-compose file yet? I can't seem to find any reference to it.

**Comment by aaronlehmann** [comment]
I don't believe it is supported in compose yet. I just filed #32584 to track this.

**Pull request merged**
